### ISSUE TITLE
Length-prefixes and flow control

### ISSUE THREAD

**Issue by martinthomson** [quicwg/base-drafts#1432]
HTTP/QUIC has a particular example of a pattern that seems likely to be usable in many other contexts.  A block of data is preceded by a length field.  The receiver of that block might read the length field then await the arrival of the entire block before processing the block.  During that time, the block is left unread so that the data can sit in the receive buffer of the transport.  This ensures that the data counts towards the flow control limits of the receiver.  This can prevent the sender from adding more messages that would only consume additional memory at the receiver.

In an extreme case, this leads to a deadlock.  The receiver is waiting for the end of a block.  The sender cannot send the remainder of the block because it does not have enough flow control credit.

This impasse doesn't naturally resolve in this case.  Worse, even if flow control might allow a single stream to make progress because the receiver ensures that it provides credit for the largest possible message or more, multiple concurrent streams could each consume a small part of the available flow control credit without allowing any single one to make progress.

Ultimately, there isn't an easy fix for this.  We could recommend a new input for prioritization: favour the sending of data for streams that are in the middle of sending a chunk, but that requires the transport be informed of message boundaries.  Relying on RST_STREAM to cancel streams until the logjam clears might work in some cases, but we have some examples already of streams that cannot be safely cancelled.  So this will always be a danger.  We gain some insight into what is going on with BLOCKED and friends, but ultimately this is just *hard*.

Not sure how much advice we can give here, or whether saying anything is even needed.

**Comment by mikkelfj** [comment]
How big is a block in this context? It can't be arbitrary because then strategy of waiting until the end before crediting flow won't work. If it isn't arbitrary, could flow be reserved for the reminder such that new streams can't be created if they cannot reserve enough flow?

**Comment by martinthomson** [comment]
This is the fun bit, I don't know.  We have frames in hq that are like this.  Some of those could be very large (think DATA), others small.  Some are naturally suited to streaming (DATA again), others not at all (HEADERS).  Smaller sizes change the odds, but they don't prevent the problem completely until the size hits zero.

I don't think that we're going to create a technical solution to this one.  I suspect that it will remain as one of the bits of the protocol that will be hard to get right.

**Comment by mikkelfj** [comment]
You could define a negotiable, but fixed, block size and require the application to drain after receiving that much, and prevent too many streams without additional block credits. This can still deadlock if the app won't buffer the block or it needs to create more streams, but it is not as arbitrary as random deadlocks.

**Comment by MikeBishop** [comment]
Because the possible sizes on HQ frames are so very large, I tend to think of them as being "regions" in the stream.  Some are have to be received entire before you can do much useful with them (HEADERS, PRIORITY); others can reasonably be streamed (DATA).  HEADERS are the easiest place to deadlock, and one of the holdovers from HTTP/2 is an explicit communication of the largest header payload you're willing to process.

**Comment by kazuho** [comment]
IMO, this is also an issue with TCP applications using _peek_ to see if a complete set of data is available in the receive buffer. The reason that it has not been considered as an issue is because TCP has expected applications to read data as they arrive, rather than using peek.

Considering that, could we recommend the same approach in the transport draft? Does doing so have a severe impact on HQ?

**Comment by MikeBishop** [comment]
The issue is cross-stream dependencies.  If you craft a very large message which requires data from another stream to process it, if the application sucks up the large message anyway, then it will explode its memory.  The point of flow control is to limit this, by constraining the *unprocessed* data at any time.  We're using unread as a stand-in for unprocessed, and the corollary is that you shouldn't read something you know you won't be able to process.  The guidance in HQ is to read far enough to decide whether you can process the message, and if not, stop reading until you can.

**Comment by janaiyengar** [comment]
What @kazuho said. This is something that an application ought to be aware of when treating QUIC's stream buffers as storage. 

The problem is caused by something broader than the structure in the application data bytes -- an application could choose to not read data from the buffer until, for instance, the next consumer is ready for it.

In all these cases, I think it is not a good idea for an app to assume that the stream's buffers are a good place to store data, since that can interfere with network throughput and with resource allocation for other streams.

**Comment by krasic** [comment]
janaiyengar@  if an app isn't to use stream buffers for storage, what does it do in your example when the next consumer isn't keeping up?   if it drains all data from the stream, it may blow up its own memory, unless it implements its own flow control protocol at the app level?   that seems.. unsatisfactory.

**Comment by dtikhonov** [comment]
Side note: the Huffman coding does have the EOS symbol.

**Comment by martinthomson** [comment]
Yeah, @krasic has it right.  The entire point of this thing is to apply back pressure - if you can't read, ensure that you don't get more of stuff you can't read that consumes more of the memory you don't have.

On the other hand, it is reasonable to say that applications (and application protocols) can't rely on QUIC providing them with any concrete amount of storage.   But we can't force people to read or we undermine the entire purpose of providing flow control.  (Here I'm assuming that reading releases flow control, but the key point is that you only release flow control once you are done with the octets, and for many of these cases, the point is that you can't be done with the octets until some external condition is met, or all the octets are in, or something like that.)

I don't think that this is something we have to provide a concrete solution for.  Flow control is fundamentally hard because it is about managing access to a finite resource and contention can occur, sometimes in sub-optimal ways.  What we can do is signpost the obvious pitfalls.  Like the case where you have 10 octets of connection flow control, 10 streams and 1 octet arrives for every stream, but you need to have >1 to make progress.

Using `peek()` is just an example, and not a necessary one.  The problem is that we're creating ways for an application to learn whether it wants to read octets before it reads them.  A length-prefixed encoding does that inherently, but it's just an example of a general class of things that a reader might use to decide to defer consumption of data.  That length-prefixes are also a great way to optimize processing of data means that we won't ever avoid using them, and nor can we avoid cases where atomic units take more than one octet to express.

If I get @dtikhonov's point, it seems like a suggestion that we might avoid a length-prefix, but I think that only makes reading harder, and doesn't avoid other styles of atomic processing over multi-octet fields.  It probably drives implementation complexity up, and it (unfortunately) uses a lot more bytes than a length prefix in many cases.

**Comment by dtikhonov** [comment]
@martinthomson, mine wasn't a suggestion, but rather something to keep in mind as we weigh different options.  That is, there is a way to not use prefix length at all.  Your analysis is correct.  The biggest downside to me is that the EOS symbol is so long.

**Comment by janaiyengar** [comment]
@krasic Yes, you're right, and I wasn't being precise. It's not always a problem, but it can become one, based on how the application manages its data.

@martinthomson: I agree that we shouldn't be looking to provide a concrete solution for, but we can generally say something about the potential implications of consuming stream buffers for data that can be read. Beyond that, I think trying to capture this application design pattern -- worthy as it is of noting somewhere -- might be reaching too far for HTTP/QUIC or QUIC docs, since it applies to not just QUIC.

**Comment by kazuho** [comment]
@MikeBishop Thank you for the clarification. I can understand the fact that we want to have a way to signal back-pressure when more than one stream is involved.

OTOH, as we all agree, using back-pressure has the danger of seeing a deadlock. It makes sense to have the implications being clarified in the documents.

And I would like to second @martinthomson's argument that the issue is not related to length-prefixes. Consider using HTTP/1 on QUIC. It makes perfect sense for an application to keep the response headers in the receive buffer until CR-LF-CR-LF arrives.

**Comment by MikeBishop** [comment]
IIRC, HPACK EOS is so long because you never actually send all of it – if the Huffman-encoded string doesn’t end on a byte boundary (and it probably doesn’t), you put as many bits of the EOS symbol as needed to reach the byte boundary.  The whole thing should never appear on the wire; at most seven bits of it.  Of course, if you’re depending on recognizing it as it goes by, that requires parsing each symbol as it passes; the length prefix makes it much easier.

**Comment by martinthomson** [comment]
(Just an aside.)  HPACK never needed EOS.  The loss of efficiency is tiny (one bit for some never-used symbol), but it should have been possible to use any sequence that is longer than the number of remaining bits.

**Comment by RyanAtGoogle** [comment]
If you want to process an entire block at once, you must have the entire block in memory, obviously. Since no application wants to buffer an unlimited amount of data, the application will have some limit on the size of the block that it accepts. If it reads the length and the length is greater than the limit, it can abort. Otherwise, I would argue, it should read the block out of the receive buffer and into an application buffer. This avoids deadlock and also prevents unlimited buffering.

**Comment by dtikhonov** [comment]
@martinthomson, @MikeBishop -- surely one can stream _HEADERS_. It's just a matter of programming.

**Comment by martinthomson** [comment]
This isn't a question of a particular construct; I'm more concerned about what advice we might give generally.  Ryan's advice is only really part of that (though it's pretty solid and pragmatic advice).

There are probably other things to say about what a sender might do that help.  For instance, if you think that something might be parsed as a single unit, then it's probably not a good idea to only send part of it.  Waiting until flow control credit is available - while it might risk a stall in the event that credits never arrive - might be a good idea.

It's also probably necessary to say something about inter-stream dependencies along the lines of: "If you have a dependency between the content of two streams, then you need to ensure that the dependent information is not sent until the information it depends on can be sent.  In particular, the necessary flow control credit needs to have been consumed."

**Comment by dtikhonov** [comment]
I think this issue should be tagged `-http` instead of `-transport`.  This logic is something for HQ to figure out.

@martinthomson writes:
> It's also probably necessary to say something about inter-stream dependencies along the lines of: "If you have a dependency between the content of two streams, then you need to ensure that the dependent information is not sent until the information it depends on can be sent. In particular, the necessary flow control credit needs to have been consumed."

IIRC, this was one of the conclusions reached by the Header Compression Design Team.  Nevertheless, I don't see any verbiage addressing this either HTTP or QPACK drafts.

**Comment by martinthomson** [comment]
We have #1420 for the specific advice, but I had hoped to gather some more general advice that that could reference.

**Comment by RyanAtGoogle** [comment]
It seems like the general advice should be something along the lines of, "applications should not use a stream's receive buffer to store data while they wait for more data to arrive, as this can potentially lead to deadlock. instead they should read data out of the buffers expeditiously."

Really receive buffers should only be used to store data when the application is temporarily blocked from processing data and should be read as soon as the application is unblocked. What do you think?

**Comment by ianswett** [comment]
It's worth a note about in-order data as well, since typically the main purpose of the receive buffer is to re-assemble data in order.

**Comment by RyanAtGoogle** [comment]
That's a good point. The receive buffer *exists* to re-assemble data in order. The receive buffer is *sized* to allow the pipe to be filled by the sender in the face of packet loss and receiver application jitter. (If a full CWND worth of data is sent, but the first packet is lost, we don't want to have to wait for the retransmission of that data to be received before we can send new data. We also don't want to have to stop sending after a full CWND just because the app got swapped out temporarily, or is waiting for a seek, for example). It does not exist as an alternative to application data storage and any attempt to treat it as such is fraught with peril.

**Comment by mirjak** [comment]
I guess it could be a good idea to give some more general advise (not only HTTP-specific) in the applicability statement. There is a section on "Use of Streams" and we could add another subsection to explain how this deadlock can happen and given some guidance inline with what @RyanAtGoogle and @martinthomson said. Anybody willing to send a PR directly or should I just open an ops-draft issue for now?

**Comment by dtikhonov** [comment]
I'd like to make another comment related to the EOS symbol.  The QPACK encoding process produces two outputs: the encoder stream instructions and the header block.  Now that the encoder framing has been removed, the encoder instructions can be written directly to the stream, while the associated header block must be buffered because we don't know its length [1,2].

We can avoid buffering of the header block by introducing an _End of the Header Block_ instruction at the cost of one byte.  The header block has just a few instructions and can easily accommodate another one.  This would also force our hand with regard to the Base Index: we'd have to pick it immediately.  On the other hand, it can make the design simpler: just set it to the same value as the Largest Reference.  This will offset the cost of the new EOS instructions, as Base Reference no longer has to be transmitted.

For this to work, we'd have to bring back _CONTINUATION_ frames, which seem to have had been designed exactly with this use case in mind.  Without them, we still have to buffer to calculate the size of the _HEADERS_ frame.  For background, see issue #76 and the [httpbis PR #548](https://github.com/http2/http2-spec/pull/548).  Needless to say, I do not agree with the rationale behind removing CONTINUATION frames from QUIC.

Footnotes:

1. I am aware of the deadlock hazard brought on by writing the header block and the encoder stream concurrently.  The assumption is that an implementation that writes to packets directly has the logic to do this right.
1. One alternative is to pre-calculate the encoded length of the header block.  This is unworkable because it a) essentially doubles the work; and b) causes buffering of unencoded header fields.

**Comment by dtikhonov** [comment]
> We can avoid buffering of the header block by introducing an _End of the Header Block_ instruction at the cost of one byte. [...]

This won't work -- the Largest Reference must still precede the header block. \*sigh\* I guess we must buffer....

**Comment by MikeBishop** [comment]
Actually, CONTINUATION frames were designed because the maximum supported size of a header set in some implementations was potentially much larger than the maximum size of a single HTTP/2 frame.  HQ eliminates them because our maximum frame size is much larger.  Sending chunks of headers at a time was a later proposal to effectively grease support for CONTINUATION, since they'd be relatively rare otherwise.

HTTP/2 still requires the entire compressed header set to be sent at once -- HEADERS+CONTINUATION effectively functions as a single jumbo frame, and didn't permit any other frame (including from other streams) to come between them.

**Comment by martinthomson** [comment]
Closing in favour of quicwg/ops-drafts#53
